kong/async.lua

```diff
@@ -0,0 +1,164 @@
+local semaphore = require "ngx.semaphore"
+
+
+local ngx = ngx
+local kong = kong
+local pcall = pcall
+local select = select
+local unpack = unpack
+local setmetatable = setmetatable
+
+
+local QUEUE_SIZE = 100000
+
+
+local function get_pending(self)
+  local head = self.head
+  local tail = self.tail
+  if head < tail then
+    head = head + QUEUE_SIZE
+  end
+  return head - tail
+end
+
+
+local function job_thread(self, index)
+  while not ngx.worker.exiting() do
+    local ok, err = self.work:wait(1)
+    if ok then
+      if self.head ~= self.tail then
+        local tail = self.tail == QUEUE_SIZE and 1 or self.tail + 1
+        local job = self.jobs[tail]
+        self.tail = tail
+        self.jobs[tail] = nil
+        ok, err = job()
+        if not ok then
+          kong.log.err("async thread #", index, " job error: ", err)
+        end
+      end
+
+    elseif err ~= "timeout" then
+      kong.log.err("async thread #", index, " wait error: ", err)
+    end
+  end
+
+  return true
+end
+
+
+local function init_worker_timer(premature, self)
+  if premature then
+    return true
+  end
+
+  local t = kong.table.new(100, 0)
+
+  for i = 1, 100 do
+    t[i] = ngx.thread.spawn(job_thread, self, i)
+  end
+
+  local ok, err = ngx.thread.wait(t[1],  t[2],  t[3],  t[4],  t[5],  t[6],  t[7],  t[8],  t[9],  t[10],
+                                  t[11], t[12], t[13], t[14], t[15], t[16], t[17], t[18], t[19], t[20],
+                                  t[21], t[22], t[23], t[24], t[25], t[26], t[27], t[28], t[29], t[30],
+                                  t[31], t[32], t[33], t[34], t[35], t[36], t[37], t[38], t[39], t[40],
+                                  t[41], t[42], t[43], t[44], t[45], t[46], t[47], t[48], t[49], t[50],
+                                  t[51], t[52], t[53], t[54], t[55], t[56], t[57], t[58], t[59], t[60],
+                                  t[61], t[62], t[63], t[64], t[65], t[66], t[67], t[68], t[69], t[70],
+                                  t[71], t[72], t[73], t[74], t[75], t[76], t[77], t[78], t[79], t[80],
+                                  t[81], t[82], t[83], t[84], t[85], t[86], t[87], t[88], t[89], t[90],
+                                  t[91], t[92], t[93], t[94], t[95], t[96], t[97], t[98], t[99], t[100])
+
+  if not ok then
+    kong.log.err("async thread error: ", err)
+  end
+
+  for i = 100, 1, -1 do
+    ngx.thread.kill(t[i])
+  end
+
+  return init_worker_timer(ngx.worker.exiting(), self)
```

Comment:
why this recursion? essentially the error above should not happen. My 2cts; if it does happen just error out and leave it there. The system will be in an undetermined state anyway since workers failed that shouldn't have failed.

Reply:
Recursion is better. In normal case we end up here when one (or more) of the threads terminates. The most common case for thread to terminate is that Nginx is shutdowning. This just calls function then once, as it checks the `premature` as a first thing, and then it terminates it. Doing recursion will not make state undetermined as we kill all the threads. The queue can still be filled while there are no threads processing them. If for some reason we did exit with some kind of error we still restart the threads. That's it. TL;DR; in 99.999% of the time it already does what you are describing.

Reply:
the not-undertermined state you are referring to is from this code perspective I think? from Kong perspective it is an undetermined state, if one of those jobs is supposed to handle some invalidation, that now doesn't run. Then there is no way to know what will go wrong next, but probably a lot. It should forcefully exit with a big error.

Reply:
All jobs will run, why wouldn't they? The `tail` is moved just before running the job.